crates/sui/src/unit_tests/cli_tests.rs

```diff
@@ -668,6 +668,48 @@ async fn test_package_publish_command_with_unpublished_dependency_fails(
     Ok(())
 }
 
+#[sim_test]
+async fn test_package_publish_command_non_zero_unpublished_dep_fails() -> Result<(), anyhow::Error>
+{
+    let with_unpublished_dependencies = true; // Value under test, results in error response for module with non-zero deps.
+
+    let mut test_cluster = TestClusterBuilder::new().build().await?;
+    let address = test_cluster.get_address_0();
+    let context = &mut test_cluster.wallet;
+
+    let client = context.get_client().await?;
+    let object_refs = client
+        .read_api()
+        .get_objects_owned_by_address(address)
+        .await?;
+
+    let gas_obj_id = object_refs.first().unwrap().object_id;
+
+    let mut package_path = PathBuf::from(TEST_DATA_DIR);
+    package_path.push("module_publish_with_unpublished_dependency_with_non_zero_address");
+    let build_config = BuildConfig::default();
+    let result = SuiClientCommands::Publish {
+        package_path,
+        build_config,
+        gas: Some(gas_obj_id),
+        gas_budget: 20_000,
+        skip_dependency_verification: false,
+        with_unpublished_dependencies,
+    }
+    .execute(context)
+    .await;
+
+    let expect = expect![[r#"
+        Err(
+            ModulePublishFailure {
+                error: "The following modules in package dependencies set a non-zero address:\nPackage UnpublishedNonZeroAddress — Module non_zero sets non-zero address 0000000000000000000000000000000000000000000000000000000000000bad\nIf these packages really are unpublished, the addresses should be set to \"0x0\" when publishing. If they are already published, ensure they specify the address in the `published-at` of the Move.toml manifest and remove --with-unpublished-dependencies from the command line command. If neither these changes are possible, take action to first publish package dependencies, or alternatively set addresses to 0x0.",
```

Comment:
This is only really instructive if run from the command line (and not, returned for `Publish` errors on the SDK path which we'll handle later). Not sure if we can make it more generic, or if we'll just special case to handle this SDK-side.

Reply:
There's no need to remove `--with-unpublished-dependencies` to get things working, the first two options you gave are sufficient, and in that case I think they will also apply to the SDK case (also some wording nits):

```suggestion
            ModulePublishFailure {
                error: "The following modules in package dependencies set a non-zero self-address:\n - 0000000000000000000000000000000000000000000000000000000000000bad::non_zero in dependency UnpublishedNonZeroAddress\nIf these packages really are unpublished, their self-addresses should be set to \"0x0\" in the [addresses] section of the manifest when publishing. If they are already published, ensure they specify the address in the `published-at` of the Move.toml manifest.",
```